pytest_rts/utils/common.py

```diff
@@ -91,6 +92,38 @@ def get_tests_from_changes(
     return tests
 
 
+def intersect_with_surroundings(changed_lines: Set[int], mapped_lines: Set[int]) -> Set[int]:
+    """
+    Finds lines existing in both changed_lines and mapped_lines.
+    For lines that exist in changed_lines but not in mapped_lines
+    looks for the closest mapped lines from left and right
+    hand sides - mapped lines which surround changed line.
+
+    Example:
+        changed_lines | 1       5             21    30
+        mapped_lines  |    2 3    10 11 12 20 21 22
+        result        |    2 3    10          21 22
+    """
+    mapped = changed_lines.intersection(mapped_lines)
+    unmapped = changed_lines.difference(mapped)
+
+    mapped_lines_sorted = sorted(mapped_lines)
+    for line in unmapped:
+        left = None
+        right = None
+        for mapped_line in mapped_lines_sorted:
```

Comment:
`O(NM)` complexity, quadratic, I wonder if we can do better. Lemme think...

Reply:
Something like this, `O(N+M)`:
```
def intersect_with_surroundings(changed_lines: Set[int], mapped_lines: Set[int]) -> Set[int]:
    """
    Return set which for every line in changed_lines contains its left and right neighbours from mapped_lines.
    
    If there is exactly the same line in mapped_lines, it is considered to be both left and right neighbour.
    
    Example:
    changed_lines = {1, 5, 21, 30}
    mapped_lines = {2, 3, 10, 11, 12, 20, 21, 22}
    
    Result: {
        2,  # right neighbour of 1
        3,  # left neighbour of 5
        10,  # right neighbour of 5
        21,  # left and reight neighbour of 21
        22  # left neighbour of 30
    }
    """
    changed_iter = sorted(changed_lines, reverse=True)
    mapped_iter = sorted(mapped_lines, reverse = True)
    res = set()
    left = None
    
    try:
        # We will be iterating in parallel sorted (in ascending order) changed and mapped lines
        cur_changed = changed_iter.pop()
        cur_mapped = mapped_iter.pop()
        
        while True:
            # The loop is terminated by IndexError exception when we run out either of changed or of mapped lines
            
            if cur_changed == cur_mapped:
                # If currently considered changed line is also mapped line, great! Add it to the result
                # and advance both cur_changed and cur_mapped.
                res.add(cur_changed)
                cur_changed = changed_iter.pop()
                cur_mapped = mapped_iter.pop()
                
            elif cur_changed < cur_mapped:
                # Current changed line is to the left of the current mapped line; in other
                # world the current mapped line is right neighbour of the current changed line.
                # Add it!
                res.add(cur_mapped)
                # It is possible that there are other changed lines such that the line we just added
                # is their right neighbour too; skip all of them.
                while cur_changed < cur_mapped:
                    cur_changed = changed_iter.pop()
                # If the changed line we've reached is the current mapped line,
                # we could advance both cursors; for simpler code we do nothing and in the next loop
                # iteration `if cur_changed == cur_mapped` condition will fire, adding the current mapped
                # line to the result again (no big deal, it's a set) and advancing the cursors.            
            else:
                # The current changed line is to the right of the current mapped line =>
                # we need to find the left neighbour of the current changed line.
                while cur_mapped <= cur_changed:
                    left = cur_mapped  # potential left neighbour
                    cur_mapped = mapped_iter.pop()
                res.add(left)
                left = None
                # We DO NOT advance current changed line, because we still need right neighbour for it.
    except IndexError:
        # We run out either of changed or of mapped lines
        if left:
            # If it happened during the search for left neighbour, we need to add the last mapped line.
            res.add(left)

    return res
```

I still need to check if it has any bugs, it's just a quick prototype.

Reply:
Please correct me if I'm wrong:
* complexity of initial algorithm is O(min(len(n), len(m)) + O(n * log(n)) + O(n * m)
* proposed algorithm is O(n * log(n)) + O(m * log(m)) + O(n + m)
* another approach could be O(n * log(n)) + O(n * log(m))

```python
def intersect_with_surroundings(changed_lines: Set[int], mapped_lines: Set[int]) -> Set[int]:
    mapped = set()
    unmapped = changed_lines.difference(mapped)
    mapped_lines_sorted = sorted(mapped_lines)
    for line in unmapped:
        left, right = get_left_right(mapped_lines_sorted, line)
        if left is not None:
            mapped.add(left)
        if right is not None:
            mapped.add(right)
    return mapped

def get_left_right(mapped_lines_sorted, line):
    if line < mapped_lines_sorted[0]:
        return None, mapped_lines_sorted[0]
    
    if line > mapped_lines_sorted[-1]:
        return None, mapped_lines_sorted[-1]

    middle = len(mapped_lines_sorted) // 2 - 1
    if (mapped_lines_sorted[middle] == line) or (mapped_lines_sorted[middle + 1] == line):
        return None, line

    if mapped_lines_sorted[middle] > line:
        return get_left_right(mapped_lines_sorted[:middle], line)

    if mapped_lines_sorted[middle + 1] < line:
        return get_left_right(mapped_lines_sorted[middle + 1:], line)

    return mapped_lines_sorted[middle], mapped_lines_sorted[middle + 1]
```

But... taking a look at what happens in line 77 with a call to [coverage_data.contexts_by_lineno(changed_file)](https://github.com/nedbat/coveragepy/blob/169972546b699f01813131d5356d0edcfc739343/coverage/sqldata.py#L911), where for every changed file a new connection to SQLite is established and numbits extraction happens, one can say improving performance of `intersect_with_surroundings()` makes no sense.

So, here, I would prefer readability (the initial version) over minor improvements in performance, however, I don't insist and can change to the proposed solution.

Reply:
Agreed, let's leave it as it is.